src/test/java/com/google/cloud/spanner/r2dbc/SpannerResultTest.java

```
@@ -17,27 +17,94 @@
+    Struct struct2 = Struct.newBuilder().set("id").to(Value.string("key2")).build();
+
+    MockResults mockResults = new MockResults();
+    mockResults.structs = Arrays.asList(struct1, struct2);
+    this.resultSet = mock(ResultSet.class);
+    when(this.resultSet.next()).thenAnswer((invocation) -> mockResults.next());
+    when(this.resultSet.getCurrentRowAsStruct())
+        .thenAnswer((invocation) -> mockResults.getCurrent());
+  }
+
+  @Test
+  public void getRowsUpdatedTest() {
+    assertThat(((Mono) new SpannerResult(this.resultSet).getRowsUpdated()).block()).isEqualTo(0);
+    assertThat(((Mono) new SpannerResult(2).getRowsUpdated()).block()).isEqualTo(2);
+  }
+
   @Test
-  public void getRowsUpdatedDummyImplementation() {
-    SpannerResult result = new SpannerResult();
-    assertThat(result.getRowsUpdated()).isNull();
+  public void nullResultSetTest() {
+    assertThatThrownBy(() -> new SpannerResult(null))
+        .hasMessage("A non-null ResultSet is required.");
   }
 
   @Test
-  public void mapDummyImplementation() {
-    SpannerResult result = new SpannerResult();
-    Flux<String> mappingResult = result.map((row, rowMetadata) -> "dummy result");
-    assertThat(mappingResult).isNotNull();
-    assertThat(mappingResult.blockFirst()).isEqualTo("dummy result");
+  public void mapTest() {
+    assertThat(new SpannerResult(this.resultSet).map((row, metadata) ->
+        ((SpannerRow) row).getStruct().getString("id") + "-" + ((SpannerRowMetadata) metadata)
+            .getStruct().getString("id")).collectList().block())
+        .containsExactly("key1-key1", "key2-key2");
+    verify(this.resultSet, times(1)).close();
   }
 
+  @Test
+  public void mapErrorTest() throws InterruptedException {
+    try {
+      new SpannerResult(this.resultSet).map((row, metadata) -> {
+        throw new RuntimeException();
+      }).onErrorStop().collectList().block();
+    } catch (RuntimeException e) {
+      verify(this.resultSet, times(1)).close();
```

This verify is likely never called. Could you check?

---

It is indeed called (breakpoint stops there), but I can move it into a finally block so it's for sure called.

---

My concern is that if some behavior in our map() implementation changes to eat the exception, this test will still pass, as resultset.close() will be called on successful completion.
Maybe add a false flag for exception thrown and set it in catch? Then verify that it's set.